### ISSUE TITLE
Implement MNIST demo

### ISSUE THREAD

**Issue by higepon** [higepon/mosh#168]
To demonstrate R7RS conformance I will write MNIST demo in R7RS.
It should work in other R7RS implementations with no or small modifications.

**Comment by higepon** [comment]
- https://github.com/higepon/mosh/pull/169
- https://github.com/higepon/mosh/pull/170

**Comment by higepon** [comment]
- https://github.com/higepon/mosh/pull/171

**Comment by higepon** [comment]
Next steps
- Try python's load-mnist
- Check dimensions of key matrix
  -  See if I have to support more than 2D metrix.

**Comment by higepon** [comment]
- https://github.com/higepon/mosh/pull/174

**Comment by higepon** [comment]
- https://github.com/higepon/mosh/pull/175

**Comment by higepon** [comment]
- https://github.com/higepon/mosh/pull/177

**Comment by higepon** [comment]
- https://github.com/higepon/mosh/pull/178

**Comment by higepon** [comment]
I want srfi-194 because I need make-normal-generator.

ToDo
- [ ] Add ```(scheme cxr)```. We don't have one!
- [ ] Add ```(scheme vectors)```
  - [ ] Check the spec.
  - [ ] Add the library.
  - [ ] Add ```(mosh chibi test)``` written in  ```(mosh test)```
  - [ ] Add the tests.
- [ ] Add srfi 133
  - [ ] Confirm ```(scheme vectors)``` == ```(srfi 1333)```
  - [ ] Add the library.
- [ ] Add ```(srfi 194)```
  - [ ]  Add the library
  - [ ]  Add the test

**Comment by higepon** [comment]
https://github.com/higepon/mosh/pull/187

**Comment by higepon** [comment]
## TODO
- ```(library (dataset mnist))```

**Comment by higepon** [comment]
Two layer neural network is really slow even for MNIST example. By running the profiler we found matrix-mul is the bottleneck.

### Observations
- We use vector of vector as 2D matrix. But it's better to use uniform vector so that we can reduce type checking.
- Calculating matrix multiplication in Scheme issues heap allocation in the most inner loop. This slows it down a lot.

### How we solve it
I'm thinking of introducing uniform 2D array with f64 type and implement it in C++. We initially support only 2D shape and f64 type.
- (make-f64-array shape fill-value)
- setter and getter.
- (matrix-mul-f64 f64-array f64-array)


More formal design comes with some iterations.
TODO: naming.

**Comment by higepon** [comment]
Maybe we use C++ template such as Array<double>, Array<int>。

**Comment by higepon** [comment]
### Naming
- SRFI 63
  - ```(make-array prototype k1 ...)```
  - ```(array-ref array k1 ...)```
  - ```(array-set! array obj k1 ...)```
- SRFI 160
  - ```(make-f64vector size [fill]) ```
  - ```(f64vector-ref @vec i) ```
  - ```(f64vector-set! @vec i value) ```
- Ours
  - ```(make-f64array fill k1 k2 ...)```
  - ```(f64-array-ref array k1 k2..)```
  - ```(f64-array-set! array value k1 k2 ...)```

**Comment by higepon** [comment]
Next steps
- See if training process runs fast enough.
  - 2-4 epochs
  - 1 epochs = 60000 / 100 = 600 steps
  - hidden_size=50
  - batch_size=100
  - learning_rate=0.1
- See if the loss goes down as expected.
- Validation in the test dataset
- Save and restore the model

**Comment by higepon** [comment]
### With native f64array implementation
2 epochs
```
loss= 2.2919106944753493
loss= 0.8038808434529071

real    1m10.895s
user    1m10.539s
sys     0m0.342s
```

**Comment by higepon** [comment]
#### ToDo
- [ ] Easy to switch to f64array implementation.
- [ ] Fix Windows build.
- [ ] Rewrite it to epoch based training.
- [ ] Show test accuracy.
- [ ] Write csv (loss and accuracy).

**Comment by higepon** [comment]
test accuracy=0.9
loss= 0.2592215852843228

**Comment by higepon** [comment]
The demo worked both on Gauche and Mosh in R7RS style.
https://github.com/higepon/mosh/tree/gosh/tests/mnist
